### DISCUSSION TITLE
[Question] How to replace only one image with AnimationImageProvider?

### DISCUSSION THREAD

**Discussion by NikKovIos** [airbnb/lottie-ios#1977] [General]
```swift
extension ViewController: AnimationImageProvider {
    func imageForAsset(asset: ImageAsset) -> CGImage? {
        guard asset.height >= 370 else {
            return nil
        }

        let image = UIImage(named: "shadow_left_ic")
        return image?.cgImage
    }
}
```

If I return nil - the image not visible. But I don't know how to map ImageAsset to CGImage. Help me please.

**Comment by calda** [comment]
Your `LottieAnimationView`'s `AnimationImageProvider` is the only data source used for images, so must return a non-nil value for any image you want to display.

If you want to customize a specific image and use the existing loading behavior for other images, you could do something like:

```swift
class MyCustomImageProvider: AnimationImageProvider {
  init(base: AnimationImageProvider?) {
    baseImageProvider = base
  }

  let baseImageProvider: AnimationImageProvider?

  func imageForAsset(asset: ImageAsset) -> CGImage? {
    guard someCondition else {
      // fall back to the base image provider for images we don't want to customize
      return baseImageProvider?.imageForAsset(asset: asset)
    }

    let image = UIImage(named: "shadow_left_ic")
    return image?.cgImage
  }
}
```

and:

```swift
let existingImageProvider = animationView.imageProvider
animationView.imageProvider = MyCustomImageProvider(base: existingImageProvider)
```

**Reply by NikKovIos** [reply]
Yes, it works, thanks!

**Reply by johnrogers** [reply]
Hello there, we are trying to follow your suggestion, albeit in SwiftUI.

We are trying to replace _only some_ images using local images, and wish to fall back to using the placeholders in the Lottie JSON files if an image with the given asset ID is not found locally.

The problem that we are facing is that initially, `animationView.imageProvider` is set to the correct base provider (the one configured by Lottie to find the placeholder images), but after setting our implementation of `AnimationImageProvider` to `animationView.imageProvider`, obviously the next time `MyCustomImageProvider` is initialised, the base provider is now going to be the same object, which causes infinite recursion.

e.g.:

```
LottieView()
    .configure { animationView in
        animationView.imageProvider = WrappedLottieImageProvider(
            imageProviderDictionary: viewStore.imageProviderDictionary,
            baseImageProvider: animationView.imageProvider // the next time this is evaluated, `baseImageProvider` becomes an instance of `WrappedLottieImageProvider`
        )
    }
```

Obviously we have to use the base provider and can't return `nil` from our `AnimationImageProvider` implementation, because otherwise it'll fail to find the placeholder images from the JSON.

Here's our implementation of our image provider:

```
struct WrappedLottieImageProvider: AnimationImageProvider {
    let imageProviderDictionary: [String: String]
    let baseImageProvider: AnimationImageProvider?
    
    init(imageProviderDictionary: [String: String], baseImageProvider: AnimationImageProvider?) {
        self.imageProviderDictionary = imageProviderDictionary
        self.baseImageProvider = baseImageProvider
    }
    
    func imageForAsset(asset: ImageAsset) -> CGImage? {
        guard
            let base64 = imageProviderDictionary[asset.id],
            let data = Data(base64Encoded: base64),
            let image = UIImage(data: data)
        else {
            return baseImageProvider?.imageForAsset(asset: asset)  // this will be `self` when the SwiftUI view is reevaluated and `animationView.imageProvider` is already `self`.
        }
        let targetWidth: CGFloat = 316
        return image.aspectFitScaleToWidth(targetWidth).cgImage
    }
}
```

Are there any ways around this? Is there any way to only set the image provider **once** when the animation changes, or get a reference to the _initial_ image provider _before_ a custom one has been set. Perhaps this could be added as a property on `LottieAnimationView`, something like `baseImageProvider` so that we can always reference it in order to fall back to placeholder images. This would be **amazing**.

Thank you for your time.
